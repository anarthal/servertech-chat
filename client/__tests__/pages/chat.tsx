import { fireEvent, render, screen } from "@testing-library/react";
import ChatPage, { ChatScreen } from "@/pages/chat";
import "@testing-library/jest-dom";
import {
  ClientMessagesEvent,
  HelloEvent,
  Room,
  ServerMessagesEvent,
  User,
} from "@/lib/apiTypes";
import WS from "jest-websocket-mock";
import userEvent from "@testing-library/user-event";
import { useRouter } from "next/router";
import { clearHasAuth } from "@/lib/hasAuth";

jest.mock("@/hooks/useIsSmallScreen");
jest.mock("@/lib/hasAuth");

describe("chat page", () => {
  // Data to mock the server
  const u1: User = { id: 1, username: "User one" };
  const u2: User = { id: 2, username: "User two" };
  const beastRoom: Room = {
    id: "beast",
    name: "Boost.Beast",
    hasMoreMessages: false,
    messages: [
      { id: "2-0", user: u2, timestamp: 1000, content: "Message 2 beast" },
      { id: "1-0", user: u1, timestamp: 900, content: "Message 1 beast" },
    ],
  };
  const wasmRoom: Room = {
    id: "wasm",
    name: "Web Assembly",
    hasMoreMessages: false,
    messages: [
      { id: "2-0", user: u2, timestamp: 12000, content: "Message 2 wasm" },
      { id: "1-0", user: u1, timestamp: 11000, content: "Message 1 wasm" },
    ],
  };
  const dbRoom: Room = {
    id: "db",
    name: "Database connectors",
    hasMoreMessages: false,
    messages: [
      { id: "2-0", user: u2, timestamp: 10100, content: "Message 2 db" },
      { id: "1-0", user: u1, timestamp: 10000, content: "Message 1 db" },
    ],
  };
  const emptyRoom: Room = {
    id: "empty",
    name: "Empty room",
    hasMoreMessages: false,
    messages: [],
  };

  // ChatScreen contains the page rendering, without state-management and
  // network interactions
  test("ChatScreen component", () => {
    // Render
    const { asFragment } = render(
      <ChatScreen
        rooms={[beastRoom, wasmRoom]}
        currentRoom={wasmRoom}
        currentUserId={2}
        onClickRoom={jest.fn()}
        onMessage={jest.fn()}
        isSmallScreen={false}
      />,
    );

    // Rooms rendered
    expect(screen.getByText("Boost.Beast")).toBeInTheDocument();
    expect(screen.getByText("Web Assembly")).toBeInTheDocument();

    // Only messages for the wasm room are rendered, since it's the one selected
    expect(screen.getByText("Message 1 wasm")).toBeInTheDocument();
    expect(screen.getAllByText("Message 2 wasm").length).toBe(2); // latest message
    expect(screen.queryByText("Message 1 beast")).toBeNull();
    expect(screen.queryByText("Message 2 beast")).toBeInTheDocument(); // latest message

    // Only user one's avatar is shown, since we're user two
    expect(screen.getByText("User one")).toBeInTheDocument();
    expect(screen.queryByText("User two")).toBeNull();

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  // ChatPage contains the entire page, including websocket interactions and
  // state management. We use an in-process, mock webserver to simulate the
  // C++ server
  describe("ChatPage", () => {
    // Data
    const helloEvt: HelloEvent = {
      type: "hello",
      payload: {
        me: { id: 10, username: "user1" },
        rooms: [beastRoom, wasmRoom, dbRoom, emptyRoom],
      },
    };

    // Cleanup any data and connections from previous tests
    beforeEach(() => {
      WS.clean();
      jest.clearAllMocks();
    });

    test("Room navigation", async () => {
      // create a WS instance, listening on the expected port
      const server = new WS(process.env.NEXT_PUBLIC_WEBSOCKET_URL);

      // render the page. This should connect the websocket
      render(<ChatPage />);
      await server.connected;

      // The page shows loading until the hello event is received
      expect(screen.getByText("Loading...")).toBeInTheDocument();

      // Hello event
      server.send(JSON.stringify(helloEvt));

      // The client draws the screen, selecting the room with the
      // latest message
      expect(screen.getByText("Boost.Beast")).toBeInTheDocument();
      expect(screen.getByText("Web Assembly")).toBeInTheDocument();
      expect(screen.getByText("Message 1 wasm")).toBeInTheDocument(); // not the latest
      expect(screen.getAllByTestId("room-name")[0]).toHaveTextContent(
        "Web Assembly",
      );

      // Clicking on a room switches to it
      fireEvent.click(screen.getByText("Database connectors"));
      expect(screen.getByText("Message 1 db")).toBeInTheDocument();
      expect(screen.queryByText("Message 1 wasm")).toBeNull();

      // Clicking on an empty room is OK
      fireEvent.click(screen.getByText("Empty room"));
      expect(screen.getByText("No more messages to show")).toBeInTheDocument();
      expect(screen.queryByText("Message 1 db")).toBeNull();
    });

    test("Sending and receiving messages", async () => {
      // Set up
      const server = new WS(process.env.NEXT_PUBLIC_WEBSOCKET_URL);

      // render the page and send the hello
      render(<ChatPage />);
      await server.connected;
      server.send(JSON.stringify(helloEvt));
      expect(screen.getAllByTestId("room-name")[0]).toHaveTextContent(
        "Web Assembly",
      ); // wasm is the room with the latest message

      // A message is received, and the target room is not selected
      const serverMsgsEvt: ServerMessagesEvent = {
        type: "serverMessages",
        payload: {
          roomId: "db",
          messages: [
            {
              id: "3-0",
              user: u1,
              timestamp: Date.now(),
              content: "A new db message",
            },
          ],
        },
      };
      server.send(JSON.stringify(serverMsgsEvt));

      // The room moves up and the message shows in the room entry
      expect(screen.getAllByTestId("room-name")[0]).toHaveTextContent(
        "Database connectors",
      );
      expect(screen.getByText("A new db message")).toBeInTheDocument();

      // A message is sent by the user. It is sent to the server
      userEvent.keyboard("A message for the wasm guys{Enter}");
      const clientMsgsEvt = JSON.parse((await server.nextMessage) as string);
      const expectedClientMsgsEvt: ClientMessagesEvent = {
        type: "clientMessages",
        payload: {
          roomId: "wasm",
          messages: [
            {
              content: "A message for the wasm guys",
            },
          ],
        },
      };
      expect(clientMsgsEvt).toStrictEqual(expectedClientMsgsEvt);

      // It doesn't appear until acked by the server
      expect(screen.queryByText("A message for the wasm guys")).toBeNull();

      // The server relays it
      const serverMsgsEvt2: ServerMessagesEvent = {
        type: "serverMessages",
        payload: {
          roomId: "wasm",
          messages: [
            {
              id: "5-0",
              user: u2,
              timestamp: Date.now(),
              content: "A message for the wasm guys",
            },
          ],
        },
      };
      server.send(JSON.stringify(serverMsgsEvt2));

      // It appears both in the room entry and in the messages screen
      expect(screen.getAllByText("A message for the wasm guys")).toHaveLength(
        2,
      );
      expect(screen.getAllByTestId("room-name")[0]).toHaveTextContent(
        "Web Assembly",
      );
    });

    test("On authentication failure, navigate to the login page", async () => {
      // Set up
      const server = new WS(process.env.NEXT_PUBLIC_WEBSOCKET_URL);

      render(<ChatPage />);
      await server.connected;
      server.close({ code: 1008, reason: "", wasClean: true });

      expect(useRouter().replace).toHaveBeenCalledTimes(1);
      expect(useRouter().replace).toHaveBeenCalledWith("/login");
      expect(clearHasAuth).toHaveBeenCalledTimes(1);
    });
  });
});
