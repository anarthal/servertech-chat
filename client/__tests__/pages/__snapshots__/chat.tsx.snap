// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat page ChatScreen component 1`] = `
<DocumentFragment>
  <div
    class="flex-1 flex min-h-0"
  >
    <div
      class="flex-1 flex flex-col overflow-y-scroll"
    >
      <div
        class="flex pt-2 pl-3 pr-3 cursor-pointer overflow-hidden"
      >
        <div
          class="flex-1 flex p-4 rounded-lg overflow-hidden"
        >
          <div
            class="pr-3 flex flex-col justify-center"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1chcy2i-MuiAvatar-root"
            >
              B
            </div>
          </div>
          <div
            class="flex-1 flex flex-col overflow-hidden"
          >
            <p
              class="m-0 font-bold pb-2"
              data-testid="room-name"
            >
              Boost.Beast
            </p>
            <p
              class="m-0 overflow-hidden text-ellipsis whitespace-nowrap"
            >
              Message 2 beast
            </p>
          </div>
        </div>
      </div>
      <div
        class="flex pt-2 pl-3 pr-3 cursor-pointer overflow-hidden"
      >
        <div
          class="flex-1 flex p-4 rounded-lg overflow-hidden"
        >
          <div
            class="pr-3 flex flex-col justify-center"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1d308z4-MuiAvatar-root"
            >
              W
            </div>
          </div>
          <div
            class="flex-1 flex flex-col overflow-hidden"
          >
            <p
              class="m-0 font-bold pb-2"
              data-testid="room-name"
            >
              Web Assembly
            </p>
            <p
              class="m-0 overflow-hidden text-ellipsis whitespace-nowrap"
            >
              Message 2 wasm
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-[3] flex flex-col"
    >
      <div
        class="flex-1 flex flex-col-reverse p-5 overflow-y-scroll"
      >
        <div
          class="flex flex-row pt-3 pb-3"
        >
          <div
            class="flex-[4]"
          />
          <div
            class="flex-[6] bg-green-100 rounded-lg pt-4 pl-5 pr-2"
          >
            <p
              class="m-0"
            >
              Message 2 wasm
            </p>
            <p
              class="m-0 pb-2 pr-2 text-xs text-right"
            >
              Jan 1, 1970, 1:00 AM
            </p>
          </div>
        </div>
        <div
          class="flex flex-row pt-3 pb-3"
        >
          <div
            class="pr-5 flex flex-col justify-end"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-vx4iuu-MuiAvatar-root"
            >
              U
            </div>
            <p
              class="m-0 pt-1 text-sm"
            >
              User one
            </p>
          </div>
          <div
            class="flex-[6] bg-white rounded-lg pt-4 pl-5 pr-2"
          >
            <p
              class="m-0"
            >
              Message 1 wasm
            </p>
            <p
              class="m-0 pb-2 pr-2 text-xs text-right"
            >
              Jan 1, 1970, 1:00 AM
            </p>
          </div>
          <div
            class="flex-[3]"
          />
        </div>
      </div>
      <div
        class="flex"
      >
        <div
          class="flex-1 flex p-2"
        >
          <input
            class="flex-1 text-xl pl-4 pr-4 pt-2 pb-2 border-0 rounded-xl"
            placeholder="Type a message..."
            type="text"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
